internal/db/copy.go

```diff
@@ -42,3 +40,71 @@ func CopyAllObjects(ctx context.Context, tx pgx.Tx, source int64, target int64)
 
 	return nil
 }
+
+func CloneToProject(ctx context.Context, tx pgx.Tx, source int64, target int64, vrange VersionRange) (*int64, error) {
+	objectQuery := &pb.ObjectQuery{
+		Path:        "",
+		IsPrefix:    true,
+		WithContent: false,
+		WithHash:    true,
+	}
+
+	latestVersion, err := LockLatestVersion(ctx, tx, target)
+
+	if err != nil {
+		return nil, fmt.Errorf("copy to project could not lock target (%d) version: %w", target, err)
+	}
+
+	newTargetVersion := latestVersion + 1
+
+	builder := newQueryBuilder(source, vrange, objectQuery)
+	removedSql, removedArgs := builder.buildRemoved()
+	removedArgsLength := len(removedArgs)
+
+	sql := fmt.Sprintf(`
+	%v
+	UPDATE dl.objects
+	SET stop_version = $%d
+	FROM removed_objects
+	WHERE "removed_objects".path = "objects".path
+	AND "objects".project = $%d
+	AND "objects".stop_version IS NULL
+	`, AsCte(removedSql, "removed_objects"), removedArgsLength+1, removedArgsLength+2)
+
+	removedArgs = append(removedArgs, newTargetVersion, target)
+
+	_, err = tx.Exec(ctx, sql, removedArgs...)
+
+	if err != nil {
+		return nil, fmt.Errorf("copy to project could not update removed files to version (%d): %w", latestVersion, err)
+	}
+
+	changedSql, changedArgs := builder.buildChanged()
+	changedArgsLength := len(changedArgs)
+
+	sql = fmt.Sprintf(`
+	%v
+	INSERT INTO dl.objects (project, start_version, stop_version, path, hash, mode, size, packed)
+	SELECT $%d, $%d, null, path, (h1, h2)::hash, mode, size, packed
+	FROM changed_objects
+	WHERE deleted IS false
+	ON CONFLICT
+	   DO NOTHING
+	`, AsCte(changedSql, "changed_objects"), changedArgsLength+1, changedArgsLength+2)
+
+	changedArgs = append(changedArgs, target, newTargetVersion)
+
+	_, err = tx.Exec(ctx, sql, changedArgs...)
+
+	if err != nil {
+		return nil, fmt.Errorf("copy to project could not insert updated files: %w", err)
+	}
+
+	err = UpdateLatestVersion(ctx, tx, target, newTargetVersion)
```

Comment:
This should also be in the `fs.go` RPC handler.

The goal is to try and avoid mixing raw SQL and the steps necessary to process an RPC.